GHzDACs/Cleanup/dac.py

```diff
@@ -936,8 +1024,534 @@ class DacRunner_Build11(DacRunner_Build7):
+
+
+# Jump table #
+
+class DacRunner_Build15(DacRunner_Build7):
+    # noinspection PyMissingConstructor
+    def __init__(self, dev, reps, start_delay, jt_entries, jt_counters, sram, loop_delay):
+        """ Make a DAC runner for a given device.
+
+        See DAC_Build15.make_jump_table for info on the jt_entries and
+        jt_counters arguments.
+
+        :param DAC_Build15 dev: the device
+        :param reps: number of times to run the board
+        :param start_delay: start delay for this board
+        :param list[(str, list[int])] jt_entries: JT entries
+        :param list[int] jt_counters: JT counter values, or None for all 0s
+        :param str sram: string data to be loaded to sram
+        :return:
+        """
+        self.dev = dev
+        self.reps = reps
+        self.start_delay = start_delay
+        self.loop_delay = loop_delay
+        self.jt_entries = jt_entries
+        self.jt_counters = jt_counters
+        self.sram = sram
+        self.nPackets = 0  # we don't expect any packets back
+        self.seqTime = fpga.TIMEOUT_FACTOR * (100 * self.reps) * 10**-6 + 1  # TODO: what should we do here? issue #49
+
+    def pageable(self):
+        return False  # no paging for JT
+
+    def loadPacket(self, page, isMaster):
+        """ Create pipelined load packet.  For DAC, upload JT and SRAM.
+
+        Note that this add 2 us to the delay for the master board.
+
+        :return: packet for the direct ethernet server
+        """
+        if isMaster:
+            # TODO: how can we add a delay to the JT?
+            self.start_delay += MASTER_SRAM_DELAY
+        jt = self.dev.make_jump_table(entries=self.jt_entries,
+                                      counters=self.jt_counters)
+        return self.dev.load(jt, self.sram)
+
+    def runPacket(self, page, slave, delay, sync):
+        """ Create run packet.
+
+        :param int page: must be 0 for JT boards
+        :param int slave: 0==master, 1==slave, 3==idle
+        :param int delay: additional delay to add to this board
+        :param int sync: passed through to sync option for register packet
+        :return: ndarray, ready to be tostring'ed to bytes for the DE server
+        """
+        start_delay = self.start_delay + delay
+        regs = self.dev.regRun(self.reps, page, slave, start_delay, readback=False,
+                               blockDelay=None, sync=sync, loop_delay=self.loop_delay)
+        return regs
+
+
+class DAC_Build15(DAC_Build8):
+    """ DAC Build 15 is the first (working) jump table build.
+
+    Note that when this DAC object builds jump tables, there are few offsets
+    that it always applies:
+
+    * Every from address has 2 subtracted to account for FPGA timing offsets.
+    * The END entry has an additional 1 subtracted.
+    * IDLE has one clock cycle subtracted from idle length as well.
+
+    A note about from addresses:
+
+    We subtract 2 from the from_addr passed into the JumpEntry. The reason for
```

Comment:
I don't understand this. It's saying that we're fussing with the from addresses to conform to some convention about where the sequence ends. But, the sequence only ends once, at one jump entry (with op code END), so why does this consideration come into *all* of the jump entries?